System/Posix/Pty.hs

```diff
@@ -280,13 +285,13 @@ foreign import capi unsafe "sys/ioctl.h value TIOCPKT_DOSTOP"
 foreign import capi unsafe "sys/ioctl.h value TIOCPKT_NOSTOP"
     tiocPktNoStop :: Word8
 
-foreign import ccall "pty_size.h"
+foreign import ccall unsafe "pty_size.h"
```

Comment:
Is there any specific reason these imports should be `unsafe`?

Reply:
Yes, for performance. Foreign code that's not expected to a) block or b) call back into Haskell code can be marked `unsafe` to reduce overhead; see https://wiki.haskell.org/Foreign_Function_Interface#Unsafe_calls

Reply:
What they forget to mention is that, even for non-blocking functions unsafe can be rather problematic if the function is slow as for the duration of the foreign call the GC is blocked. So if GC is triggered all threads will be blocked and waiting idle for the duration of the foreign call.

So I'd be hesitant to change that, *especially* for foreign code making syscalls (like these) since the overhead of a safe call is tiny compared to the overhead of making a syscall.

Reply:
> the overhead of a safe call is tiny compared to the overhead of making a syscall

That's not what I thought, do you have a citation? From what I can google the overhead of `safe` is pretty significant and can make the overall call 10x slower; see [here](https://gist.github.com/nh2/8508695). Quoting [Simon Marlow](http://blog.ezyang.com/2010/07/safety-first-ffi-and-threading/): "safe calls are more expensive and involve taking and releasing locks, and possibly creating a new OS thread."  I'm sure it depends on the syscall as well. For the relatively simple calls like get/set_pty_size I'd expect `unsafe` to be a win.

> So if GC is triggered all threads will be blocked and waiting idle for the duration of the foreign call.

Not sure what you mean by this -- I believe GHC will simply not trigger GC while an `unsafe` call is active, so there's no scenario where other capabilities not involved in the `unsafe` call get blocked. Given the syscall duration is probably measured in 10s/100s of nanoseconds and typical GC frequency is measured in 100s of milliseconds it should not be a problem.

Reply:
Of course safe calls are more expensive. The question is "does that matter?". You're kinda glossing over the takeaway right at the top of Edward's post :)
> Don’t use unsafe in your FFI imports! We really mean it!

Anyway, I'm fairly familiar with the low level details of safe vs unsafe calls and I'm not really inclined to mark foreign calls unsafe unless I'm 100% it's 1) necessary and 2) safe. So, in the absence of data showing a problem I'll keep them as is. I'll merge this without them, do a bit of local testing and then make a new release on Hackage.

Reply:
> You're kinda glossing over the takeaway right at the top of Edward's post

Not really, he acknowledges in the comments that the C code he's dealing with is doing intensive things like solving SAT problems. I still don't think this rule applies to a short C function with a single syscall / don't get why you'd rather trade off a guaranteed performance hit on every call to avoid delaying a GC iteration by a tiny amount (unless I'm wrong about the "blocking every thread" scenario?). But agreed it's unlikely to cause a noticeable problem.

Thanks for merging!